### DISCUSSION TITLE
Newbie question

### DISCUSSION THREAD

**Discussion by kesavkolla** [bazelbuild/rules_rust#1913] [Q&A]
I have multi crate repo where one crate is depends on another. Here is high level structure of my repo:

├── BUILD.bazel
├── Cargo.lock
├── Cargo.toml
├── config
├── muspell-server.iml
├── README.md
├── src
│   ├── binaries
│   │   ├── cache
│   │   │   ├── Cargo.toml
│   │   │   └── src
│   │   │       └── main.rs
│   │   └── server
│   │       ├── BUILD.bazel
│   │       ├── Cargo.toml
│   │       └── src
│   │           └── main.rs
│   ├── common
│   │   ├── BUILD.bazel
│   │   ├── Cargo.toml
│   │   └── src
│   │       └── lib.rs
│   └── query
│       ├── analytics
│       │   ├── Cargo.toml
│       │   └── src
│       │       └── lib.rs
│       ├── postgres
│       │   ├── Cargo.toml
│       │   └── src
│       │       └── lib.rs
│       ├── redis
│       │   ├── Cargo.toml
│       │   └── src
│       │       └── lib.rs
│       └── resources
│           ├── Cargo.toml
│           └── src
│               └── lib.rs
└── WORKSPACE

My binaries use the other crates as dependencies. When I tried with simple Bazel.build file the binary is not building. It didn't pickup it's own Cargo.toml for dependencies.

Here is my BUILD.bazel for one of the binary:

```bazel
package(default_visibility = ["//visibility:public"])

load("@rules_rust//rust:defs.bzl", "rust_binary")

rust_binary(
    name = "server",
    srcs = ["src/main.rs"],
    deps = ["//src/common"],
)

```

I get compilation error 

```
error[E0433]: failed to resolve: use of undeclared crate or module `tokio`
 --> src/binaries/server/src/main.rs:1:3
  |
1 | #[tokio::main]
  |   ^^^^^ use of undeclared crate or module `tokio`

error[E0752]: `main` function is not allowed to be `async`
 --> src/binaries/server/src/main.rs:2:1
  |
2 | async fn main() {
  | ^^^^^^^^^^^^^^^ `main` function is not allowed to be `async`

error: aborting due to 2 previous errors
```

Can someone points me where I was doing wrong?

**Comment by illicitonion** [comment]
Bazel doesn't implicitly add dependencies to your BUILD files, you need to tell Bazel that you want your dependencies to be added. There's a macro to request "all dependencies from the adjacent Cargo.toml file" you can use to make this easier.

See https://github.com/bazelbuild/rules_rust/blob/4663ff6a3d0c2abb79cf98dca269503364be921b/examples/crate_universe/cargo_local/BUILD.bazel#L9-L10 for an example.

**Comment by kesavkolla** [comment]
Thanks for pointing me. My curious question: What is crate_universe means?

I would appreciate if someone who is maintainer of this repo to publish a nice how-to blog post about this. This is an awesome project and rust ecosystem is growing. I am sure there are lot of folks who are in same boat as me scratching head on how to start.

**Reply by illicitonion** [reply]
crate_universe is the support in rules_rust for pulling in crates.io dependencies to build with Bazel. There's some documentation in https://bazelbuild.github.io/rules_rust/crate_universe.html and some examples in https://github.com/bazelbuild/rules_rust/tree/main/examples/crate_universe but as you note, no how-to style post at the moment.
